src/items/recipes.zig

```diff
@@ -0,0 +1,214 @@
+const std = @import("std");
+const main = @import("main");
+const items = main.items;
+const ZonElement = main.ZonElement;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const NeverFailingArenaAllocator = main.heap.NeverFailingArenaAllocator;
+const ItemStack = items.ItemStack;
+const Tag = main.Tag;
+const Recipe = items.Recipe;
+const BaseItemIndex = items.BaseItemIndex;
+const Block = main.blocks.Block;
+
+const Segment = union(enum) {literal: []const u8, symbol: []const u8};
+
+fn parsePattern(allocator: NeverFailingAllocator, pattern: []const u8) !main.List(Segment) {
+	var arenaAllocator: NeverFailingArenaAllocator = .init(main.stackAllocator);
+	defer arenaAllocator.deinit();
+	const arena = arenaAllocator.allocator();
+	var segments: main.List(Segment) = .init(arena);
+	var idx: usize = 0;
+	while(idx < pattern.len) {
+		if(pattern[idx] == '{') {
+			idx += 1;
+			const endIndex = std.mem.indexOfScalarPos(u8, pattern, idx, '}') orelse return error.UnclosedBrackets;
+			if(idx == endIndex) return error.EmptyBrackets;
+			const symbol = pattern[idx..endIndex];
+			segments.append(.{.symbol = symbol});
+			idx = endIndex + 1;
+		} else {
+			const endIndex = std.mem.indexOfScalarPos(u8, pattern, idx, '{') orelse pattern.len;
+			segments.append(.{.literal = pattern[idx..endIndex]});
+			idx = endIndex;
+		}
+	}
+	var newSegments: main.List(Segment) = .init(allocator);
+	for(segments.items) |segment| {
+		switch(segment) {
+			.literal => |literal| {
+				newSegments.append(.{.literal = allocator.dupe(u8, literal)});
```

Comment:
Since these are substrings from the string from the recipe zon, which outlives all of the code here, I don't think we need to dupe this.